design-documents/web-api-controllers.md

```diff
@@ -0,0 +1,289 @@
+All of these problems stem from having business layer contracts/entities exposed as presentation layer operations/data.
+ 
+ 
+## GraphQL
+#### Current situation
+GraphQL (sometimes) utilizes service contracts and relies on them to have execute complex validations if any needed.
+It also uses simple validation provided by graphql schema (like a property not allowed to be null). However if an
+endpoint requires complex validation/authorization that is not performed by a service contract (and it shouldn't since
+service contracts must not be aware of context) resolvers would duplicate to match functionality provided by regular
+controllers providing HTML presentation. Furthermore since it's easier to define simple validation rules via graphQL
+schema GraphQL gateway may provide more proper user input validation than HTML/REST/SOAP gateways.
+ 
+ 
+## HTML Presentation
+#### Current situation
+Controllers implementing _ActionInterface_ do complex validation/authorization themselves which is fine as they are
+supposed to be responsible for user input processing, still there's a problem with other areas not being able to reuse
+this logic.
+ 
+ 
+## What needs to be done to solve this?
+#### Using operation-specific DTOs
+In order to make it more clear what data will be actually used/allowed for a certain operation and to avoid transferring
+of extensive data sets we should move from having generic DTOs describing entities to operation-specific DTOs describing
+arguments and results of an operation.
+ 
+Example on how would we change DTOs for operations regarding customer creation/update:
+ 
+First we would have DTO that describes a customer in the way it's convenient for client code when updating a customer:
+```php
+interface CustomerUpdateInterface
+{
+    //No getId() since storefront clients must not be able to control their customer's ID
+    ...
+    
+    //For setting new password
+    public function getPassword(): ?string;
+    
+    public function getAddresses(): CustomerAddress[];
+}
+```
+ 
+Then we would have DTO for customer's addresses:
+```php
+interface CustomerAddressUpdateInterface
+{
+    public function getId(): ?string;
+    ....
+    
+    //No getCustomerId() because we don't give control over that
+    ....
+    
+    public function getPostcode(): string
+}
+```
+
+Since we don't want original password in customer responses we would not return it as read operations result:
+```php
+interface CustomerReadInterface
+{
+    //Now we have getId() since it's OK for storefront clients to know their customer's ID.
+    public function getId(): string;
+    
+    ....
+    //No getPassword()
+    ....
+}
+```
+ 
+By having these operation-specific DTOs we would only have the data we need as arguments and responses.
+ 
+###### How to introduce it to Magento?
+There are couple of ways we can introduce this concept to Magento:
+* use it for new storefront API
```

Comment:
that's just one suggestion meaning that we should employ this concept when designing new storefront API since they will be new and it will be easier to design them with this concept in mind